workflows/bot: allow maintainer merges after committer approval

wolfgangwalther opened the pull request:
*(only the last commit belongs to this PR, everything else is part of #457652)*

This allows committers to approve PRs with additional, optional nits that the author-maintainer can either address or merge immediately without these changes.

It also allows committers to approve a PR for merge, while still waiting for other maintainers to give their feedback - they can then merge the PR directly instead of passing it back to the committer.

Resolves https://github.com/NixOS/nixpkgs-merge-bot/issues/232

## Things done

- [x] [Approvals from nixpkgs-core](https://github.com/NixOS/nixpkgs/blob/master/ci/README.md#nixpkgs-merge-bot), because this increases the scope of the merge-bot.
- [x] Very basic testing locally.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflows/bot: allow maintainer merges after committer approval

This allows committers to approve PRs with additional, optional nits
that the author-maintainer can either address or merge immediately
without these changes.

It also allows committers to approve a PR for merge, while still waiting
for other maintainers to give their feedback - they can then merge the
PR directly instead of passing it back to the committer.

MattSturgeon commented on ci/github-script/merge.js:
```diff
@@ -31,6 +37,19 @@ async function runChecklist({ github, context, pull_request, maintainers }) {
+    ({ event, state, commit_id }) =>
+      event === 'reviewed' &&
+      state === 'approved' &&
+      commit_id === pull_request.head.sha,
```
(Maybe this subtle but important detail deserves a comment?)

MattSturgeon commented on ci/README.md:
```diff
@@ -43,8 +43,10 @@ These issues effectively list PRs the merge bot has interacted with.
-- The PR is authored by [@r-ryantm](https://nix-community.github.io/nixpkgs-update/r-ryantm/) or a [committer][@NixOS/nixpkgs-committers].
+- The PR only touches packages located under `pkgs/by-name/*`.
+- The PR is either:
+  - approved or authored by a [committer][@NixOS/nixpkgs-committers], or
```
```suggestion
  - approved by a [committer][@NixOS/nixpkgs-committers],
  - authored by a [committer][@NixOS/nixpkgs-committers], or
```

MattSturgeon replied:
Also: should it explicitly mention "the latest push is approved by a committer"?

wolfgangwalther replied:
> Also: should it explicitly mention "the latest push is approved by a committer"?

I did not include that in the README. I think it's too much detail for what that tries to achieve. Feedback will be given, because the label will be removed once the approval doesn't count anymore - so people will be able to figure this out on their own.

MattSturgeon reviewed (approved):
LGTM, pending nixpkgs-core approval

wolfgangwalther commented:
The @NixOS/nixpkgs-core approved the change to allow maintainers merge PRs to their packages after committer approval as long as the approval was for the latest state of the PR. In general, pushes to the PR that result in a different SHA should invalidate previous approvals, although this approval covers a theoretical future extension to consider "pure rebases, which neither change the diff, nor the commit messages" as well.

(this extension is hypothetical, I have no plans to do actually do that ...)

wolfgangwalther committed:
ci/github-script/merge: restructure head SHA check

While it was already the case that only merge comments *after* the
latest push were acted on, the logic wasn't easy to understand. This
change should make it more obvious, specially in combination with the
next commit, that all steps (comments, approvals, merge) must happen on
the same SHA - the current head SHA of the PR.

wolfgangwalther commented:
> as long as the approval was for the latest state of the PR. In general, pushes to the PR that result in a different SHA should invalidate previous approvals

While this was the case before already, I found the logic hard to understand. I added a commit before the main commit to refactor this a bit. I think it is now much clearer that everything compares to `pull_request.head.sha`, which is a single source of truth. This means approval and merge *need* to happen on the same commit.

@MattSturgeon could you have another look to confirm?

dotlambda commented on ci/github-script/merge.js:
```diff
@@ -23,11 +24,25 @@ function runChecklist({
+          state === 'approved' &&
+          // Only approvals for the current head SHA count, otherwise authors could push
+          // bad code between the approval and the merge.
+          commit_id === pull_request.head.sha,
```
Shouldn't we also check that other properties like the target branch were not changed? Maybe even the PR title since that appears in the commit message.

wolfgangwalther replied:
In general the conditions to make a PR eligible for the merge-bot (automated update, backport, committer creator/approval) are not about "this PR is 100% correct down to the last dot". If they were - we could merge immediately in these cases instead of letting the maintainer merge.

This is about making sure that there is no *actively harmful code introduced* and to catch the *really* bad mistakes. Changing the PR title is certainly not that. You could maybe merge a stdenv rebuild into master by changing the target branch, but it seems like this should be solved on a different level anyway (aka preventing it for anyone, essentially).

TLDR: Maintainers need to be allowed judgement and be trusted to a certain degree as well.
